antsibull/build_ansible_commands.py

```diff
@@ -164,28 +202,36 @@ def build_single_command():
         print(f'{app_ctx.extra["build_file"]} is for version {build_ansible_version} but we need'
               f' {app_ctx.extra["ansible_version"].major}'
               f'.{app_ctx.extra["ansible_version"].minor}')
-        return 1
+        return None
 
     with tempfile.TemporaryDirectory() as tmp_dir:
         download_dir = os.path.join(tmp_dir, 'collections')
         os.mkdir(download_dir, mode=0o700)
 
         # Download included collections
+        versions: t.Optional[t.Mapping[str, SemVer]] = None
+        if dependency_data is not None:
+            versions = {
+                collection: SemVer(version)
+                for collection, version in dependency_data.deps.items()
+            }
         included_versions = asyncio.run(download_collections(deps, app_ctx.galaxy_url,
                                                              download_dir,
-                                                             app_ctx.extra['collection_cache']))
+                                                             app_ctx.extra['collection_cache'],
+                                                             force_versions=versions))
 
-        new_dependencies = DependencyFileData(
-            str(app_ctx.extra["ansible_version"]),
-            str(ansible_base_version),
-            {collection: str(version) for collection, version in included_versions.items()})
+        if dependency_data is None:
+            dependency_data = DependencyFileData(
+                str(app_ctx.extra["ansible_version_version"]),
+                str(ansible_base_version),
+                {collection: str(version) for collection, version in included_versions.items()})
```

Comment:
This code makes me think maybe it's time to refactor download_collections and what it calls.  Step (1) retrieve a mapping of collections: versions.  [which we skip if the mapping is passed in].  Step (2) download the collections at those versions.

Reply:
I've now implemented that in fa1955ef4ba1fcbea7b2c7e7a700c93f926e44a8 and 8bd8e1bdfe256b8975ce02fdbb1484a0632f2052.